src/belgium/create_db.Rmd

```diff
@@ -0,0 +1,413 @@
+  occ <- occ_download_get(
+    key = "0044021-181108115102211", 
+    path = here("data", "raw")
+  )
+}
+```
+
+We unzip text file with occurrences as  `key_number` + `occurrence.txt` in `./data/raw`:
+
+```{r unzip_csv}
+occ_file <- paste(key, "occurrence.txt", sep = "_")
+occ_path <- here("data", "raw", occ_file)
+
+if (!file.exists(here("data", "raw", occ_file))) {
+  unzip(zipfile = occ,
+        files = "occurrence.txt",
+        exdir = here("data", "raw"))
+  file.rename(from = here("data", "raw", "occurrence.txt"),
+              to = occ_path
+  )
+}
+``` 
+
+Name of columns:
+
+```{r get_cols}
+cols_occ_file <- read_delim(
+  occ_path, "\t", n_max = 1,
+  quote = ""
+)
+cols_occ_file <- names(cols_occ_file)
+```
+
+Number of columns present:
+
+```{r n_cols}
+length(cols_occ_file)
+```
+
+## Create sqlite file
+
+### Database name, path and table name
+
+Define name and path of `.sqlite` file:
+
+```{r sqlite_path}
+sqlite_file <- paste(key, "occurrence.sqlite", sep = "_")
+sqlite_path <- here("data", "interim", sqlite_file)
+```
+
+and table name:
+
+```{r define_table_name}
+table_name <- "occ_be_all"
+```
+
+### Define storage class for each column
+
+The standard storage class is `TEXT`:
+
+```{r def_default_cols_type}
+# default type: TEXT
+field_types <- rep("TEXT", length(cols_occ_file))
+names(field_types) <- cols_occ_file
+```
+
+The following columns should be of storage class `INTEGER`:
+
+1. `*Key`, e.g. `taxonKey`, `speciesKey`)
+2. `*DayOfYear`: `startDayOfYear` and  `endDayOfYear`  
+3. `year`
+4. `month`
+5. `day`
+
+```{r set_to_INTEGER}
+int_fields <- names(field_types)[str_detect(names(field_types), "Key") & 
+                                   names(field_types) != "datasetKey"]
+int_fields <- c(
+  int_fields,
+  names(field_types)[str_detect(names(field_types), "DayOfYear")],
+  names(field_types)[names(field_types) == "year"],
+  names(field_types)[names(field_types) == "month"],
+  names(field_types)[names(field_types) == "day"]
+)
+field_types[which(names(field_types) %in% int_fields)] <- "INTEGER"
+```
+
+The following columns should be of storage class `REAL`:
+
+1. `decimal*`: `decimalLatitude` and `decimalLongitude`
+2. `coordinate*`: `coordinateUncertaintyInMeters` and `coordinatePrecision`
+3. `pointRadiusSpatialFit`
+
+```{r}
+real_fields <- names(field_types)[str_detect(names(field_types), "decimal")]
+real_fields <- c(
+  real_fields,
+  names(field_types)[str_detect(names(field_types), "coordinate")],
+  names(field_types)[names(field_types) == "pointRadiusSpatialFit"]
+)
+field_types[which(names(field_types) %in% real_fields)] <- "REAL"
+```
+
+Open connection to database:
+
+```{r open_connection_to_db}
+sqlite_occ <- dbConnect(SQLite(), dbname = sqlite_path)
+```
+
+And fill it with occurrences from text file. This step reads the large occurrence file in chunks and transfer them in the sqlite file. This step can take long the first timeyou run it:
+
+```{r fill_sqlite_file}
```

Comment:
I got the error:

>  Error in connection_import_file(conn@ptr, name, value, sep, eol, skip) : 
  RS_sqlite_import: /Users/peter_desmet/Coding/Repositories/trias-project/occ-processing/data/raw/0044021-181108115102211_occurrence.txt line 3637202 expected 237 columns of data but found 226

Reply:
I run the pipeline again and I got no errors back. 
Thanks to the error report I could extract the lines around 3637202 with Windows-based text editor called [LogExpert](http://www.log-expert.de/)). I extracted the lines from 3637200 to 3637205 included (= 6 lines) and saved in [data/interim/bug_PR_6_discussion_r266808445.txt](https://github.com/trias-project/occ-processing/blob/assign-grid-belgium/data/interim/bug_PR_6_discussion_r266808445.txt). 
I don't see any particular difference in line 637202. They are all machine observations of gulls (_Larus_).
 I would first try running the pipeline again :crossed_fingers: 
If still not working, try to read the file I mentioned above in R. 
If nothing anomalous found, we can further check in your local MAC OS X settings.

Reply:
Can you also run the commando `devtools::session_info()` in R and let me know which version of RSQLite you have installed on your Mac? Thanks.

Reply:
Same bug.

<details>
<summary>Session info:</summary>

```r
> devtools::session_info()
─ Session info ────────────────────────────────────────────────────────
 setting  value                       
 version  R version 3.5.2 (2018-12-20)
 os       macOS Mojave 10.14.3        
 system   x86_64, darwin15.6.0        
 ui       RStudio                     
 language (EN)                        
 collate  en_US.UTF-8                 
 ctype    en_US.UTF-8                 
 tz       Europe/Brussels             
 date     2019-04-12                  

─ Packages ────────────────────────────────────────────────────────────
 package     * version    date       lib source                    
 assertthat    0.2.1      2019-03-21 [1] CRAN (R 3.5.2)            
 backports     1.1.3      2018-12-14 [1] CRAN (R 3.5.0)            
 bit           1.1-14     2018-05-29 [1] CRAN (R 3.5.0)            
 bit64         0.9-7      2017-05-08 [1] CRAN (R 3.5.0)            
 blob          1.1.1      2018-03-25 [1] CRAN (R 3.5.0)            
 broom         0.5.1      2018-12-05 [1] CRAN (R 3.5.0)            
 callr         3.1.1      2018-12-21 [1] CRAN (R 3.5.0)            
 cellranger    1.1.0      2016-07-27 [1] CRAN (R 3.5.0)            
 cli           1.1.0      2019-03-19 [1] CRAN (R 3.5.2)            
 colorspace    1.4-0      2019-01-13 [1] CRAN (R 3.5.2)            
 crayon        1.3.4      2017-09-16 [1] CRAN (R 3.5.0)            
 data.table    1.12.0     2019-01-13 [1] CRAN (R 3.5.2)            
 DBI           1.0.0.9001 2019-04-12 [1] Github (r-dbi/DBI@069df06)
 desc          1.2.0      2018-05-01 [1] CRAN (R 3.5.0)            
 devtools      2.0.1      2018-10-26 [1] CRAN (R 3.5.1)            
 digest        0.6.18     2018-10-10 [1] CRAN (R 3.5.0)            
 dplyr       * 0.8.0.1    2019-02-15 [1] CRAN (R 3.5.2)            
 forcats     * 0.3.0      2018-02-19 [1] CRAN (R 3.5.0)            
 fs            1.2.6      2018-08-23 [1] CRAN (R 3.5.0)            
 generics      0.0.2      2018-11-29 [1] CRAN (R 3.5.0)            
 geoaxe        0.1.0      2016-02-19 [1] CRAN (R 3.5.0)            
 ggplot2     * 3.1.0      2018-10-25 [1] CRAN (R 3.5.0)            
 glue        * 1.3.1      2019-03-12 [1] CRAN (R 3.5.2)            
 gtable        0.2.0      2016-02-26 [1] CRAN (R 3.5.0)            
 haven         2.0.0      2018-11-22 [1] CRAN (R 3.5.0)            
 here        * 0.1        2017-05-28 [1] CRAN (R 3.5.0)            
 hms           0.4.2      2018-03-10 [1] CRAN (R 3.5.0)            
 httr          1.4.0      2018-12-11 [1] CRAN (R 3.5.1)            
 jsonlite      1.6        2018-12-07 [1] CRAN (R 3.5.0)            
 knitr         1.21       2018-12-10 [1] CRAN (R 3.5.1)            
 lattice       0.20-38    2018-11-04 [1] CRAN (R 3.5.2)            
 lazyeval      0.2.1      2017-10-29 [1] CRAN (R 3.5.0)            
 lubridate     1.7.4      2018-04-11 [1] CRAN (R 3.5.0)            
 magrittr      1.5        2014-11-22 [1] CRAN (R 3.5.0)            
 memoise       1.1.0      2017-04-21 [1] CRAN (R 3.5.0)            
 modelr        0.1.2      2018-05-11 [1] CRAN (R 3.5.0)            
 munsell       0.5.0      2018-06-12 [1] CRAN (R 3.5.0)            
 nlme          3.1-137    2018-04-07 [1] CRAN (R 3.5.2)            
 oai           0.2.2      2016-11-24 [1] CRAN (R 3.5.0)            
 pillar        1.3.1      2018-12-15 [1] CRAN (R 3.5.0)            
 pkgbuild      1.0.2      2018-10-16 [1] CRAN (R 3.5.0)            
 pkgconfig     2.0.2      2018-08-16 [1] CRAN (R 3.5.0)            
 pkgload       1.0.2      2018-10-29 [1] CRAN (R 3.5.0)            
 plyr          1.8.4      2016-06-08 [1] CRAN (R 3.5.0)            
 prettyunits   1.0.2      2015-07-13 [1] CRAN (R 3.5.0)            
 processx      3.2.1      2018-12-05 [1] CRAN (R 3.5.0)            
 ps            1.3.0      2018-12-21 [1] CRAN (R 3.5.0)            
 purrr       * 0.3.2      2019-03-15 [1] CRAN (R 3.5.2)            
 R6            2.4.0      2019-02-14 [1] CRAN (R 3.5.2)            
 Rcpp          1.0.1      2019-03-17 [1] CRAN (R 3.5.2)            
 readr       * 1.3.1      2018-12-21 [1] CRAN (R 3.5.0)            
 readxl        1.1.0      2018-04-20 [1] CRAN (R 3.5.0)            
 remotes       2.0.2      2018-10-30 [1] CRAN (R 3.5.0)            
 rgbif       * 1.1.0.9311 2019-02-26 [1] local                     
 rgeos         0.4-2      2018-11-08 [1] CRAN (R 3.5.0)            
 rlang         0.3.4      2019-04-07 [1] CRAN (R 3.5.2)            
 rprojroot     1.3-2      2018-01-03 [1] CRAN (R 3.5.0)            
 RSQLite     * 2.1.1      2018-05-06 [1] CRAN (R 3.5.0)            
 rstudioapi    0.8        2018-10-02 [1] CRAN (R 3.5.0)            
 rvest         0.3.2      2016-06-17 [1] CRAN (R 3.5.0)            
 scales        1.0.0      2018-08-09 [1] CRAN (R 3.5.0)            
 sessioninfo   1.1.1      2018-11-05 [1] CRAN (R 3.5.0)            
 sp            1.3-1      2018-06-05 [1] CRAN (R 3.5.0)            
 stringi       1.4.3      2019-03-12 [1] CRAN (R 3.5.2)            
 stringr     * 1.4.0      2019-02-10 [1] CRAN (R 3.5.2)            
 testthat      2.0.1      2018-10-13 [1] CRAN (R 3.5.0)            
 tibble      * 2.1.1      2019-03-16 [1] CRAN (R 3.5.2)            
 tidyr       * 0.8.2      2018-10-28 [1] CRAN (R 3.5.0)            
 tidyselect    0.2.5      2018-10-11 [1] CRAN (R 3.5.0)            
 tidyverse   * 1.2.1      2017-11-14 [1] CRAN (R 3.5.0)            
 usethis       1.4.0      2018-08-14 [1] CRAN (R 3.5.0)            
 whisker       0.3-2      2013-04-28 [1] CRAN (R 3.5.0)            
 withr         2.1.2      2018-03-15 [1] CRAN (R 3.5.0)            
 xfun          0.4        2018-10-23 [1] CRAN (R 3.5.0)            
 xml2          1.2.0      2018-01-24 [1] CRAN (R 3.5.0)            
 yaml          2.2.0      2018-07-25 [1] CRAN (R 3.5.0)            

[1] /Library/Frameworks/R.framework/Versions/3.5/Resources/library
> 
```
</details>

Reply:
Ok, my file is truncated when unzipping, which caused the file to abruptly end at 3637202:

```r
In unzip(zipfile = download_file, files = "occurrence.txt", exdir = here("data",  :
  possible truncation of >= 4GB file
```

Reply:
It did work by using `unzip filename` in bash.